src/field_theory/minpoly.lean

```diff
@@ -386,38 +386,94 @@ by simpa [sub_eq_add_neg] using add_algebra_map hx (-a)
+over the fraction field. See `minpoly.gcd_domain_eq_field_fractions'` if `S` is already a
+`K`-algebra. -/
+lemma gcd_domain_eq_field_fractions :
+  minpoly K (algebra_map S L s) = (minpoly R s).map (algebra_map R K) :=
 begin
-  symmetry,
-  refine eq_of_irreducible_of_monic _ _ _,
+  refine (eq_of_irreducible_of_monic _ _ _).symm,
   { exact (polynomial.is_primitive.irreducible_iff_irreducible_map_fraction_map
-      (polynomial.monic.is_primitive (monic hx))).1 (irreducible hx) },
-  { have htower := is_scalar_tower.aeval_apply A K R x (minpoly A x),
-    rwa [aeval, eq_comm] at htower },
-  { exact (monic hx).map _ }
+      (polynomial.monic.is_primitive (monic hs))).1 (irreducible hs) },
+   { rw [aeval_map, aeval_def, is_scalar_tower.algebra_map_eq R S L, ← eval₂_map, eval₂_at_apply,
+      eval_map, ← aeval_def, aeval, map_zero] },
+  { exact (monic hs).map _ }
+end
+
+/-- For GCD domains, the minimal polynomial over the ring is the same as the minimal polynomial
+over the fraction field. Compared to `minpoly.gcd_domain_eq_field_fractions`, this version is useful
+if the element is in a ring that is already a `K`-algebra. -/
+lemma gcd_domain_eq_field_fractions' [algebra K S] [is_scalar_tower R K S] :
+  minpoly K s = (minpoly R s).map (algebra_map R K) :=
+begin
+  let L := fraction_ring S,
+  rw [← gcd_domain_eq_field_fractions K L hs],
+  refine minpoly.eq_of_algebra_map_eq (is_fraction_ring.injective S L)
+    (is_integral_of_is_scalar_tower _ hs) rfl
 end
 
 /-- For GCD domains, the minimal polynomial divides any primitive polynomial that has the integral
 element as root. -/
-lemma gcd_domain_dvd {A R : Type*} (K : Type*)
-  [comm_ring A] [is_domain A] [normalized_gcd_monoid A] [field K]
-  [comm_ring R] [is_domain R] [algebra A K]
-  [is_fraction_ring A K] [algebra K R] [algebra A R] [is_scalar_tower A K R]
-  {x : R} (hx : is_integral A x)
-  {P : A[X]} (hprim : is_primitive P) (hroot : polynomial.aeval x P = 0) :
-  minpoly A x ∣ P :=
+lemma gcd_domain_dvd (hinj : function.injective (algebra_map R S))
+  {P : R[X]} (hP : P ≠ 0) (hroot : polynomial.aeval s P = 0) : minpoly R s ∣ P :=
 begin
-  apply (is_primitive.dvd_iff_fraction_map_dvd_fraction_map K
-    (monic.is_primitive (monic hx)) hprim).2,
-  rw ← gcd_domain_eq_field_fractions K hx,
+  let K := fraction_ring R,
+  let L := fraction_ring S,
+  haveI : no_zero_smul_divisors R L,
+  { refine no_zero_smul_divisors.iff_algebra_map_injective.2 _,
+    rw [is_scalar_tower.algebra_map_eq R S L],
+    exact (is_fraction_ring.injective S L).comp hinj },
+  let P₁ := P.prim_part,
+  suffices : minpoly R s ∣ P₁,
+  { exact dvd_trans this (prim_part_dvd _) },
+  have hP₁ : polynomial.aeval s P₁ = 0,
+  { rw [eq_C_content_mul_prim_part P, map_mul, aeval_C] at hroot,
+    have hcont : P.content ≠ 0 := λ h, hP (content_eq_zero_iff.1 h),
+    replace hcont := function.injective.ne hinj hcont,
+    rw [map_zero] at hcont,
+    exact eq_zero_of_ne_zero_of_mul_left_eq_zero hcont hroot },
```

Comment:
This deserves its own lemma, of the form `no_zero_smul_divisors → p ≠ 0 → aeval x p = 0 → aeval x p.prim_part = 0` (and the same for `eval₂ f` given `injective f`).